### PR TITLE
torchdynamo: add convolution pointwise(binary) fusion kernel

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86583
* __->__ #86582
* #86581

Support binary fusion of Convolution with:

- add
- sub
- mul
- div

### CODE REVIEW

**Comment by EikanWang** [comment]
@jgong5 , any comments for this PR?

File: test/test_mkldnn_fusion.py [hunk 1]
```diff
@@ -246,5 +255,40 @@ def forward(self, x):
...
+        input_shapes = {2: (112, 112), 3: (55, 55, 55)}
+        for pointwise_name, pointwise_fn in self._binary_list().items():
+            for dim in [2, 3]:
+                channels_last = torch.channels_last if dim == 2 else torch.channels_last_3d
+                options = itertools.product([True, False], [1, 2], [1, 4], [torch.contiguous_format, channels_last])
+                for bias, dilation, groups, memory_format in options:
+                    oC = 10 * groups
+                    iC = 3 * groups
+                    x_shape = (1, iC) + input_shapes[dim]
+                    x = torch.randn(x_shape, dtype=torch.float32).to(memory_format=memory_format)
```
Comment (jgong5) [inline test/test_mkldnn_fusion.py:279]: Test bfloat16 too? Same comments for other tests.

File: test/test_mkldnn_fusion.py [hunk 2]
```diff
@@ -246,5 +255,40 @@ def forward(self, x):
...
+                    mod = M(pointwise_fn, dim, iC, oC, dilation, groups, bias, kernel_size=3)
+                    mod = mod.to(memory_format=memory_format).eval()
+                    other = torch.randn_like(mod.conv(x))
```
Comment (jgong5) [inline test/test_mkldnn_fusion.py:282]: Test same dtype and different dtype?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -296,6 +296,116 @@ Tensor mkldnn_convolution_pointwise(
...
+  check_shape_forward(
+      input_t, weight_t, bias, padding, stride, dilation, groups);
+  auto output_sizes = conv_output_size(
+      input_t.sizes(), weight_t.sizes(), padding, stride, dilation);
+  // TODO: support broadcast binary fusion.
+  TORCH_CHECK(
+      output_sizes == other_t.sizes(),
+      "Binary Fusion's add should have same shape");
+  bool can_be_fused = input_t.scalar_type() == other_t.scalar_type() &&
+      input_t.device() == other_t.device() && groups == 1 &&
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Conv.cpp:324]: Yes, add device check.

File: test/test_mkldnn_fusion.py (hunk 1 above)
Comment (XiaobingSuper) [reply test/test_mkldnn_fusion.py:279]: I think we can add the bfloat16 test case at another PR which includes linear, Conv.

File: test/test_mkldnn_fusion.py (hunk 2 above)
Comment (XiaobingSuper) [reply test/test_mkldnn_fusion.py:282]: Now, I add a dtype check which makes sure all input should have same type. and for mixed-dtype, we will report an error to user.

**Review by jgong5 — APPROVED** [review]

**Comment by EikanWang** [comment]
@jansel , could you help review this PR for TorchInductor?

**Comment by EikanWang** [comment]
@XiaobingSuper , please fix the CI issue.

**Comment by XiaobingSuper** [comment]
> @XiaobingSuper , please fix the CI issue.

The error is not related to this PR.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**